### PR TITLE
Add support for encoding and decoding MessagePack

### PR DESCRIPTION
**Pull request by hughsie**
MessagePack seems to be a less good CBOR, but is used by some hardware.

I've not added support for signed ints or extensions as they're not common and not required for this hardware.

See-also: https://msgpack.org/

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: libfwupdplugin/fu-msgpack.c [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	g_autoptr(GPtrArray) items = g_ptr_array_new_with_free_func((GDestroyNotify)g_object_unref);
+
+	g_return_val_if_fail(buf != NULL, NULL);
+	g_return_val_if_fail(error == NULL || *error == NULL, NULL);
+
+	while (offset < buf->len) {
+		g_autoptr(FuMsgpackItem) item = NULL;
+		item = fu_msgpack_item_parse(buf, &offset, error);
+		if (item == NULL)
+			return NULL;
```
Comment (superm1) [inline libfwupdplugin/fu-msgpack.c:38]: not sure how common this is going to be; but if you think it's more than once it happens it's probably worth prefixing the error to indicate which message from the buffer failed to make it easier to debug in the future.

Feel free to ignore!

File: libfwupdplugin/fu-msgpack.c [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * Since: 2.0.0
+ **/
+GByteArray *
+fu_msgpack_write(GPtrArray *items, GError **error)
+{
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+
+	g_return_val_if_fail(items != NULL, NULL);
+	g_return_val_if_fail(error == NULL || *error == NULL, NULL);
```
Comment (superm1) [inline libfwupdplugin/fu-msgpack.c:62]: why can't a user call this with a NULL error?

**Review by superm1 — DISMISSED** [review]

File: libfwupdplugin/fu-msgpack.c (hunk 1 above)
Comment (hughsie) [reply libfwupdplugin/fu-msgpack.c:38]: Not going to ignore, it's a good suggestion. Including the offset would make debugging much easier.  Fixed.

File: libfwupdplugin/fu-msgpack.c (hunk 2 above)
Comment (hughsie) [reply libfwupdplugin/fu-msgpack.c:62]: It's the same mental gymnastics as the other callers, essentially what it's saying is that a NULL error is okay, and a GError ** pointing to a NULL GError * is okay, but what's not okay is if the *error is non-NULL -- which is practice means "return with NULL if the GError is already set" which stops us getting the "The GError was overwritten with a new value" critical error on the new error path -- i.e. failing fast if we forget to return from a previous error with failure.

**Pull request merged**